### PR TITLE
Move Restart logic to instance structs

### PR DESCRIPTION
**Pull request by dlemel8**

### CODE REVIEW

File: lxd/instance/drivers/driver_lxc.go [hunk 1]
```diff
@@ -2786,6 +2786,25 @@ func (c *lxc) Shutdown(timeout time.Duration) error {
 	return nil
 }
 
+// Restart restart the instance.
+func (c *lxc) Restart(timeout time.Duration) error {
+	if timeout == 0 {
+		if err := c.Stop(false); err != nil {
```
Comment (stgraber) [inline lxd/instance/drivers/driver_lxc.go:2792]: LXD coding style is to stay away from the one line defintion+check syntax.
So this should be kept as it was before on two lines, same for the ones below.

File: lxd/instance/drivers/driver_qemu.go [hunk 2]
```diff
@@ -629,6 +629,25 @@ func (vm *qemu) Shutdown(timeout time.Duration) error {
 	return nil
 }
 
+// Restart restart the instance.
+func (vm *qemu) Restart(timeout time.Duration) error {
+	if timeout == 0 {
+		if err := vm.Stop(false); err != nil {
```
Comment (stgraber) [inline lxd/instance/drivers/driver_qemu.go:635]: Same here wrt coding style

**Comment by tomponline** [comment]
@dlemel8 @stgraber the two `Restart()` functions look identical from what I can see (albeit using the respective driver vars).

How about we move this into a function called `restart(inst instance.Instance, timeout time.Duration) error` in `lxd/instance/drivers/load.go` that accepts the instance being restarted, that would then allow us to create a small `Restart(timeout time.Duration) error` function in `lxd/instance/drivers/driver_common.go` which could then call the `restart()` function we created. In this way we would avoid the duplication of logic.

**Comment by stgraber** [comment]
They won't remain identical for long as we'll be adding lifecycle events to it next

**Comment by stgraber** [comment]
(and those are type specific)

**Comment by tomponline** [comment]
@stgraber could they be derived from the instance type, so that the code itself is common?

**Comment by tomponline** [comment]
@stgraber and will a single lifecycle event be generated for "restarted" or will it be a "start" and then "stop"? If the former then the specific lifecycle event could be sent from the instance's `Restart()` function after calling the common function, and if the latter, should we be generating those events inside the associated `Start` and `Stop` functions instead?

**Comment by dlemel8** [comment]
@stgraber do you want me to change this PR according to @tomponline suggestion?

**Comment by stgraber** [comment]
@tomponline single event.

**Comment by stgraber** [comment]
@tomponline I don't think that what you're describing can work.
`driver_common.go` is of type `common` which is a subset of `Instance`.

As a result functions inside of `common` will never pass a struct which satisfies `Instance`.
We need code in lxc and qemu to be able to pass in a full `Instance` struct that we can call Stop, Shutdown and Start on.

**Comment by stgraber** [comment]
So something like this should be about as minimal as we can make it:
https://paste.ubuntu.com/p/zsXbqN44WS/

@dlemel8 I'd say update your branch to that ^ and we should be good

**Comment by tomponline** [comment]
> So something like this should be about as minimal as we can make it:
> https://paste.ubuntu.com/p/zsXbqN44WS/
> 
> @dlemel8 I'd say update your branch to that ^ and we should be good

That's great, I prefer that to having a separate common function in load.go. even if it does mean having an arg for Instance .

**Comment by stgraber** [comment]
jenkins: test this please

**Pull request merged**